src/service-api.ts

```diff
@@ -0,0 +1,292 @@
+    const children = await readdir(folder);
+
+    // we will flatMap with promises: first map
+    const uploads = children.map(async (child) => {
+      const childPath = path.join(folder, child);
+      const childUploadPath = path.join(uploadPath, child);
+      const stats = await lstat(childPath);
+
+      if (stats.isDirectory()) {
+        // recursively upload child folder
+        return await uploadH5P(childPath, childUploadPath, member);
+      } else {
+        // ignore this file if extension is not allowed
+        const ext = path.extname(childPath);
+        if (!h5pValidator.isExtensionAllowed(ext)) {
+          // we're using flatMap, represent none value with empty array
+          return [];
+        }
+
+        const mimetype = mime.getType(ext) ?? DEFAULT_MIME_TYPE;
+        const size = stats.size;
+        const uploadTask = fileTaskManager.createUploadFileTask(member, {
+          file: fs.createReadStream(childPath),
+          filepath: childUploadPath,
+          size,
+          mimetype,
+        });
+
+        await taskRunner.runSingle(uploadTask);
+        // we're using flatMap, wrap result in array
+        return [childUploadPath];
+      }
+    });
+    // then resolve promises array and flatten
+    return (await Promise.all(uploads)).flat();
+  }
+
+  /**
+   * Helper to create H5P extra
+   */
+  function buildH5PExtra(contentId: string, filename: string): H5PExtra {
+    return {
+      h5p: {
+        contentId,
+        h5pFilePath: buildH5PPath(contentId, filename), // <contentId>/<filename>.h5p
+        contentFilePath: buildContentPath(contentId), // <contentId>/content
+      },
+    };
+  }
+
+  /**
+   * Creates a Graasp item for the uploaded H5P package
+   * @param filename Name of the original H5P file WITHOUT EXTENSION
+   * @param contentId Storage ID of the remote content
+   * @param remoteRootPath Root path on the remote storage
+   * @param member Actor member
+   * @param parentId Optional parent id of the newly created item
+   */
+  async function createH5PItem(
+    filename: string,
+    contentId: string,
+    member: Actor,
+    parentId?: string,
+  ): Promise<Item<H5PExtra>> {
+    const metadata = {
+      name: buildH5PPath('', filename),
+      type: H5P_ITEM_TYPE,
+      extra: buildH5PExtra(contentId, filename),
+    };
+    const create = itemTaskManager.createCreateTaskSequence(member, metadata, parentId);
+    return taskRunner.runSingleSequence(create) as Promise<Item<H5PExtra>>;
+  }
+
+  /*
+    we create an artificial plugin scope, so that fastify-multipart does not conflict
+    with other instances since we use fp to remove the outer scope
+  */
+  await fastify.register(async (fastify) => {
+    fastify.register(fastifyMultipart, {
+      limits: {
+        fields: MAX_NON_FILE_FIELDS,
+        files: MAX_FILES,
+        fileSize: MAX_FILE_SIZE,
+      },
+    });
+
+    fastify.post<{ Querystring: { parentId?: string } }>(
+      '/h5p-import',
```

Comment:
Do we not factor out these routes?

Reply:
I don't think we have factored out any route in the back-end. This may change in the future if we decide to refactor routes into a `RouteBuilder` in the SDK.